src/number_theory/multiplicity.lean

```diff
@@ -0,0 +1,473 @@
+  { exact λ h, hx (hp.dvd_of_dvd_pow h) }
+end
+
+end integral_domain
+
+section lifting_the_exponent
+variables (hp : nat.prime p) (hp1 : odd p)
+include hp hp1
+
+/-- **Lifting the exponent lemma** for odd primes. -/
+lemma int.pow_sub_pow {x y : ℤ} (hxy : ↑p ∣ x - y) (hx : ¬↑p ∣ x) (n : ℕ) :
+  multiplicity ↑p (x ^ n - y ^ n) = multiplicity ↑p (x - y) + multiplicity p n :=
+begin
+  cases n,
+  { simp only [multiplicity.zero, add_top, pow_zero, sub_self] },
+  have h : (multiplicity _ _).dom := finite_nat_iff.mpr ⟨hp.ne_one, n.succ_pos⟩,
+  rcases eq_coe_iff.mp (enat.coe_get h).symm with ⟨⟨k, hk⟩, hpn⟩,
+  conv_lhs { rw [hk, pow_mul, pow_mul] },
+  rw nat.prime_iff_prime_int at hp,
+  rw ←int.nat_cast_eq_coe_nat at *,
+  rw [pow_sub_pow_of_prime hp, pow_prime_pow_sub_pow_prime_pow hp hp1 hxy hx, enat.coe_get],
+  { rw ←geom_sum₂_mul,
+    exact dvd_mul_of_dvd_right hxy _ },
+  { exact λ h, hx (hp.dvd_of_dvd_pow h) },
+  { iterate 2 { rw int.nat_cast_eq_coe_nat },
+    rw int.coe_nat_dvd,
+    rintro ⟨c, rfl⟩,
+    refine hpn ⟨c, _⟩,
+    rwa [pow_succ', mul_assoc] }
+end
+
+
+lemma int.pow_add_pow {x y : ℤ} (hxy : ↑p ∣ x + y) (hx : ¬↑p ∣ x) {n : ℕ} (hn : odd n) :
+  multiplicity ↑p (x ^ n + y ^ n) = multiplicity ↑p (x + y) + multiplicity p n :=
+begin
+  rw ←sub_neg_eq_add at hxy,
+  rw [←sub_neg_eq_add, ←sub_neg_eq_add, ←nat.odd.neg_pow hn],
+  exact int.pow_sub_pow hp hp1 hxy hx n
+end
+
+lemma nat.pow_sub_pow {x y : ℕ} (hxy : p ∣ x - y) (hx : ¬p ∣ x) (n : ℕ) :
+  multiplicity p (x ^ n - y ^ n) = multiplicity p (x - y) + multiplicity p n :=
+begin
+  obtain hyx | hyx := le_total y x,
+ { iterate 2 { rw ← int.coe_nat_multiplicity },
+    rw int.coe_nat_sub (nat.pow_le_pow_of_le_left hyx n),
+    rw ← int.coe_nat_dvd at hxy hx,
+    push_cast at *,
+    exact int.pow_sub_pow hp hp1 hxy hx n },
+  { simp only [nat.sub_eq_zero_iff_le.mpr hyx,
+      nat.sub_eq_zero_iff_le.mpr (nat.pow_le_pow_of_le_left hyx n), multiplicity.zero,
+        enat.top_add] }
+end
+
+lemma nat.pow_add_pow {x y : ℕ} (hxy : p ∣ x + y) (hx : ¬p ∣ x) {n : ℕ} (hn : odd n) :
+  multiplicity p (x ^ n + y ^ n) = multiplicity p (x + y) + multiplicity p n :=
+begin
+  iterate 2 { rw [←int.coe_nat_multiplicity] },
+  rw ←int.coe_nat_dvd at hxy hx,
+  push_cast at *,
+  exact int.pow_add_pow hp hp1 hxy hx hn
+end
+
+end lifting_the_exponent
+end multiplicity
+end comm_ring
+
+namespace multiplicity
+
+lemma int.pow_two_sub_pow_two_eq_factorisation {x y : ℤ} (n : ℕ) :
+  x ^ (2 ^ n) - y ^ (2 ^ n) = (∏ i in finset.range n, (x ^ (2 ^ i) + y ^ (2 ^ i))) * (x - y) :=
+begin
+  induction n with d hd,
+  { simp only [pow_zero, pow_one, finset.range_zero, finset.prod_empty, one_mul] },
+  { suffices : x ^ 2 ^ d.succ - y ^ 2 ^ d.succ = (x ^ 2 ^ d + y ^ 2 ^ d) * (x ^ 2 ^ d - y ^ 2 ^ d),
+    { rw [this, hd, finset.prod_range_succ, ← mul_assoc, mul_comm (x ^ 2 ^ d + y ^ 2 ^ d)] },
+    { ring_exp_eq }}
+end
+
+lemma int.sq_mod_four_eq_one_of_odd {x : ℤ} : odd x → x ^ 2 % 4 = 1 :=
```

Comment:
Should this one go `data.zmod.basic` instead?

Reply:
> Without the `_root_` prefix, this would end up in the `multiplicity` namespace.

I think the namespace is closed here, but I agree it's better to use `_root_`.

Reply:
> Should this one go `data.zmod.basic` instead? Also, it could be upgraded to an `iff` statment

An `iff` sounds good. As for the location, maybe in `data.int.modeq`?